File: js_checker.js
const fs = require('fs');
const path = require('path');

function checkJsFiles(dir) {
    const files = [];

    function findFiles(directory) {
        const items = fs.readdirSync(directory);
        for (const item of items) {
            const fullPath = path.join(directory, item);
            if (fs.statSync(fullPath).isDirectory()) {
                findFiles(fullPath);
            } else if (fullPath.endsWith('.js') || fullPath.endsWith('.jsx')) {
                files.push(fullPath);
            }
        }
    }
    findFiles(dir);

    for (const file of files) {
        const content = fs.readFileSync(file, 'utf8');
        const lines = content.split('\n');
        
        for (let i = 0; i < lines.length; i++) {
            const line = lines[i];
            
            // Match pure functions / components defined with const name = (...) =>
            const constMatch = line.match(/^\s*(export\s+)?const\s+([A-Za-z0-9_]+)\s*=\s*(async\s+)?\(/);
            // Match hook components or regular function
            const varMatch = line.match(/^\s*(export\s+)?const\s+([A-Za-z0-9_]+)\s*=\s*/);
            
            const funcMatch = line.match(/^\s*(export\s+)?(async\s+)?function\s+([A-Za-z0-9_]+)\s*\(/);
            
            let name = null;
            if (constMatch) {
               name = constMatch[2];
            } else if (funcMatch) {
               name = funcMatch[3];
            }

            if (name) {
                // Check previous 1-5 lines for JSDoc
                let hasJSDoc = false;
                for (let j = Math.max(0, i-5); j < i; j++) {
                    if (lines[j].includes('*/')) {
                        hasJSDoc = true;
                        break;
                    }
                }
                if (!hasJSDoc) {
                    console.log(`Missing JSDoc: ${file}:${i+1} : ${name}`);
                }
            }
        }
    }
}

checkJsFiles(process.argv[2]);

File: js_checker.py
import os
import sys
import re

def check_js_files(directory):
    missing = []
    
    # regexes
    const_re = re.compile(r'^\s*(export\s+)?const\s+([A-Za-z0-9_]+)\s*=\s*(async\s+)?\(')
    func_re = re.compile(r'^\s*(export\s+)?(async\s+)?function\s+([A-Za-z0-9_]+)\s*\(')
    
    for root, _, files in os.walk(directory):
        for file in files:
            if file.endswith('.js') or file.endswith('.jsx'):
                path = os.path.join(root, file)
                with open(path, 'r', encoding='utf-8') as f:
                    lines = f.readlines()
                    
                for i, line in enumerate(lines):
                    name = None
                    m1 = const_re.match(line)
                    if m1:
                        name = m1.group(2)
                    else:
                        m2 = func_re.match(line)
                        if m2:
                            name = m2.group(3)
                            
                    if name:
                        # check if there's JSDoc above
                        has_doc = False
                        start = max(0, i - 10)
                        for j in range(start, i):
                            if '*/' in lines[j]:
                                has_doc = True
                                break
                        if not has_doc:
                            missing.append(f"{path}:{i+1} : {name}")
    return missing

if __name__ == '__main__':
    js_dir = sys.argv[1]
    missing_docs = check_js_files(js_dir)
    if missing_docs:
        for m in missing_docs:
            print(f"Missing JSDoc: {m}")
    else:
        print("All JS functions have JSDoc!")
